aidial_sdk/application.py

```diff
@@ -139,99 +104,130 @@ def configure_telemetry(self, config: TelemetryConfig):
 
         init_telemetry(app=self, config=config)
 
-    def add_embeddings(self, deployment_name: str, impl: Embeddings) -> None:
-        self.embeddings_impls[deployment_name] = impl
+    def add_embeddings(
+        self, deployment_name: str, impl: Embeddings
+    ) -> "DIALApp":
+        self.add_api_route(
+            f"/openai/deployments/{deployment_name}/embeddings",
+            self._embeddings(deployment_name, impl),
+            methods=["POST"],
+        )
+
+        return self
 
     def add_chat_completion(
         self, deployment_name: str, impl: ChatCompletion
-    ) -> None:
-        self.chat_completion_impls[deployment_name] = impl
+    ) -> "DIALApp":
 
-    def _chat_completion_endpoint_factory(
-        self, endpoint: str, request_type: Type["RequestType"]
-    ):
-        async def _handler(
-            deployment_id: str, original_request: Request
-        ) -> Response:
-            set_log_deployment(deployment_id)
-            deployment = self._get_chat_completion(deployment_id)
+        self.add_api_route(
+            f"/openai/deployments/{deployment_name}/chat/completions",
+            self._chat_completion(deployment_name, impl),
+            methods=["POST"],
+        )
 
-            request = await request_type.from_request(original_request)
+        self.add_api_route(
+            f"/openai/deployments/{deployment_name}/rate",
+            self._rate_response(deployment_name, impl),
+            methods=["POST"],
+        )
 
-            endpoint_impl = getattr(deployment, endpoint, None)
-            if not endpoint_impl:
-                raise missing_endpoint_error(endpoint)
+        if endpoint_impl := get_method_implementation(impl, "tokenize"):
+            self.add_api_route(
+                f"/openai/deployments/{deployment_name}/tokenize",
+                self._endpoint_factory(
+                    deployment_name, endpoint_impl, "tokenize", TokenizeRequest
+                ),
+                methods=["POST"],
+            )
+
+        if endpoint_impl := get_method_implementation(impl, "truncate_prompt"):
+            self.add_api_route(
+                f"/openai/deployments/{deployment_name}/truncate_prompt",
+                self._endpoint_factory(
+                    deployment_name,
+                    endpoint_impl,
+                    "truncate_prompt",
+                    TruncatePromptRequest,
+                ),
+                methods=["POST"],
+            )
+
+        return self
+
+    def _endpoint_factory(
+        self,
+        deployment_id: str,
+        endpoint_impl: Callable[[RequestType], Coroutine[Any, Any, Any]],
+        endpoint: Literal["tokenize", "truncate_prompt"],
```

Comment:
I think we can remove this parameter and calculate it inside the function basis on `request_type`.